prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -5245,6 +5245,27 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         Ok(stmts)
     }
 
+    fn encode_array_types(&self, array_ty: ty::Ty<'tcx>) -> EncodingResult<(String, vir::Type, String, vir::Type, vir::Type, ty::Ty<'tcx>, usize)> {
+        // type predicates
+        let array_pred = self.encoder.encode_type_predicate_use(array_ty)?;
+        let (elem_ty_rs, len) = if let ty::TyKind::Array(elem_ty, len) = array_ty.kind() {
+            (elem_ty, len)
+        } else {
+            unreachable!()
+        };
+        let elem_pred = self.encoder.encode_type_predicate_use(elem_ty_rs)?;
+
+        // types
+        let array_ty = self.encoder.encode_type(array_ty)?;
+        let elem_ty = self.encoder.encode_type(elem_ty_rs)?;
+        let elem_value_ty = self.encoder.encode_value_type(elem_ty_rs)?;
+
+        let array_len = self.encoder.const_eval_intlike(&len.val)?
+            .to_u64().unwrap().try_into().unwrap();
+
+        Ok((array_pred, array_ty, elem_pred, elem_ty, elem_value_ty, elem_ty_rs, array_len))
```

Comment:
Could you define a `struct EncodedArrayTypes { array_pred: ... }` private to `procedure_encoder.rs` to avoid returning the large tuple?

Reply:
I'm going to merge this because overall it's an improvement. You can add the `EncodedArrayTypes` in a new PR.

Reply:
oh, thought about that as well while writing it, would have added in this PR if it was still open :-) but new PR is fine as well.